### PR TITLE
Port the Browser to U2

### PR DESCRIPTION
**Pull request by shepheb**
Includes:
- Porting `StackView`, `BrowserView`, `UpdateDetailView` and more.
- Fixes several contextualizing bugs in U2 infrastructure.
- Adds CSS shorthand support to the U2 element parser.
- Adds min/preferred/max widths to `Element` (used by `StackView`).

### CODE REVIEW

File: js/foam/browser/u2/StackView.js [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    'as stack'
+  ],
+
+  constants: {
+    VIEW_DESTROYED: ['view-destroyed']
+  },
+
+  properties: [
+    {
+      model_: 'IntProperty',
```
Comment (kgrgreer) [inline js/foam/browser/u2/StackView.js]: Use new style for new code:

type: 'Int'

File: js/foam/u2/UpdateView.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+CLASS({
+  package: 'foam.u2',
+  name: 'UpdateView',
+  extends: 'foam.u2.View',
```
Comment (kgrgreer) [inline js/foam/u2/UpdateView.js]: Does this need to be a View?

File: js/foam/u2/md/ToolbarAction.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+CLASS({
+  package: 'foam.u2.md',
+  name: 'ToolbarAction',
```
Comment (kgrgreer) [inline js/foam/u2/md/ToolbarAction.js]: Why is this needed?  Is this going to make configuring toolbars with templates difficult?

File: js/foam/u2/md/WithToolbar.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+CLASS({
+  package: 'foam.u2.md',
+  name: 'WithToolbar',
```
Comment (kgrgreer) [inline js/foam/u2/md/WithToolbar.js]: Include at least one line of documentation to say what this model does.

File: js/foam/u2/md/WithToolbar.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    },
+    {
+      name: 'toolbar_',
+      factory: function() {
+        var t = this.Toolbar.create({ title$: this.title$ });
+        t.addLeftActions([this.backAction_]);
+        return t;
+      }
+    },
+    ['addPassthrough_', false]
```
Comment (kgrgreer) [inline js/foam/u2/md/WithToolbar.js]: What does this do?

File: js/foam/u2/md/WithToolbar.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  methods: [
+    function initE() {
+      this.addPassthrough_ = true;
+      this.cls(this.myCls()).add(this.toolbar_).add(this.body_);
+      this.addPassthrough_ = false;
+    },
+
+    function add() {
+      if (this.addPassthrough_) return this.SUPER.apply(this, arguments);
+      else this.body_.add.apply(this.body_, arguments);
```
Comment (kgrgreer) [inline js/foam/u2/md/WithToolbar.js]: Why no return value?

I don't like this style of formatting.

File: js/foam/u2/Element.js [hunk 7]
```diff
@@ -305,7 +306,49 @@ CLASS({
       name: 'innerHTML',
       transient: true,
       getter: function() { return this.outputInnerHTML(this.createOutputStream()).toString(); }
-    }
+    },
+    {
+      name: 'minWidth',
```
Comment (adamvy) [inline js/foam/u2/Element.js]: Are these the sorts of things we should be adding to element?  It feels out of place, this is not used on all elements, only in a few context.

What about putting these in a trait only applied to views that are intended to be used in the stackview or something.

File: js/foam/browser/u2/StackView.js (hunk 1 above)
Comment (shepheb) [inline js/foam/browser/u2/StackView.js]: Done.

File: js/foam/u2/UpdateView.js (hunk 2 above)
Comment (shepheb) [inline js/foam/u2/UpdateView.js]: Maybe not? It really wants to be a transparent controller, like we discussed yesterday. For now, it is an `Element` and has `data`, so it seemed natural to just make it a `View`.

When we have a way to do transparent borders or controllers nicely, then I'll port it to that.

File: js/foam/u2/md/ToolbarAction.js (hunk 3 above)
Comment (shepheb) [inline js/foam/u2/md/ToolbarAction.js]: In U1 it was needed to associate actions that are being sent to a distant toolbar with the data they were supposed to operate on.

In U2, it probably won't be needed soon, based on the toolbar design discussion I had with Mark yesterday. We're evolving a new plan for toolbars in MD apps, and I think we will avoid needing anything like this. This (and to a lesser extent `WithToolbar`) are stepping stones, porting the U1 functionality to U2. They're not the final state.

File: js/foam/u2/md/WithToolbar.js (hunk 5 above)
Comment (shepheb) [inline js/foam/u2/md/WithToolbar.js]: It's needed for the hack of adding things to a child view. We need to discuss this further today, it's an important use case that's going to keep coming up.

File: js/foam/u2/md/WithToolbar.js (hunk 6 above)
Comment (shepheb) [inline js/foam/u2/md/WithToolbar.js]: Reformatted, and fixed the return value. for the latter case.

File: js/foam/u2/Element.js (hunk 7 above)
Comment (shepheb) [inline js/foam/u2/Element.js]: If it were in a trait that only some views supported, then you have to know about that trait and make sure you have it set on anything you want to stack. That's possible, of course, but I think it's a needless complication. `Element`, like U1 `View`, contains a lot of things that are only needed sometimes.

On the other hand, Mark and I were talking about different styles of stacking and transitions anyway, so it may be that the current stacking style will change anyway, and these become unnecessary.

File: js/foam/u2/md/ToolbarAction.js (hunk 3 above)
Comment (adamvy) [inline js/foam/u2/md/ToolbarAction.js]: I think we should try to avoid putting known throwaway code in the foam.u2 package.

If its temporary put it in another package,.

File: js/foam/u2/UpdateView.js (hunk 2 above)
Comment (adamvy) [inline js/foam/u2/UpdateView.js]: The question of why does it need to be a view is broader.  Why does it need to live in the View tree at all?

You 're decorating the view to apply some sort of model saving logic.  But why not just decorate the model (MVC) instead?  This is a model concern in MVC, not a view concern.

Using something like the foam.persistence.ObjectReplicator you apply it to your model itself, and it gives you a new model which will auto-save to a dao when changed.

Comment (jacksonic) [inline js/foam/u2/UpdateView.js]: Having the functionality implemented to attach to the model, I agree, but
also (perhaps in a separate package) it's convenient to have a view that
just does the right thing so you can define a UI and saving behaviour
without having to fiddle with your data.

We're entering EasyDAO (EasyView?) territory. It would certanily make
UpdateDetailView much smaller :P

Comment (adamvy) [inline js/foam/u2/UpdateView.js]: UpdateDetailView has visual features.

It adds save and cancel buttons.  This decorator does not, and therefore shouldn't be a view, there is nothing visual to it, it's purely part of the model layer.

Comment (shepheb) [inline js/foam/u2/UpdateView.js]: The plan for the replacement of `UpdateDetailView` is actually to default to auto-saving for now. Even if there are save/cancel actions, volunteering them to a toolbar is a separate concern. So I agree that the replacement for `UDV` need not be a `View`.

I looked at `ObjectReplicator` and it seems like it should work. I think the plan is as follows: use `ObjectReplicator` or a new, similar model to wrap the data and auto-save it. Then we construct a `DetailView` normally and pass it the replicator as its `data`. We might need a U2 `FutureView`  for that to work nicely.

Sounds like something for the next weekly design meeting.

**Comment by shepheb** [comment]
PTAL. I modernized all the MD views to use dynamic instead of on$, and enableCls where appropriate.

File: index.js [hunk 8]
```diff
@@ -106,6 +107,10 @@
       } else if (  ( X.lookup('foam.ui.View').isInstance(obj) )
                    || ( 'CView' in GLOBAL && CView.isInstance(obj) ) ) {
         view = obj;
+      } else if ( X.lookup('foam.u2.Element').isInstance(obj) ) {
+        document.body.insertAdjacentHTML('beforeend', obj.outerHTML);
+        obj.load();
+        return;
       } else if ( obj.toView_ ) {
```
Comment (adamvy) [inline index.js]: Would be nice to add a case here if obj.toE is a thing, use that.

File: js/foam/browser/BrowserConfig.js [hunk 9]
```diff
@@ -154,6 +161,11 @@ CLASS({
     },
     {
       model_: 'ViewFactoryProperty',
+      name: 'rowView',
+      defaultValue: 'foam.ui.md.CitationView',
```
Comment (adamvy) [inline js/foam/browser/BrowserConfig.js]: Shouldn't this be foam.u2.md.CitationView ?

Comment (adamvy) [inline js/foam/browser/BrowserConfig.js]: And same for detailView below.

File: js/foam/browser/u2/BrowserView.js [hunk 10]
```diff
@@ -0,0 +1,450 @@
...
+        },
+      ],
+    }
+  ],
+
+  properties: [
+    {
+      name: 'touchManager',
+      factory: function() {
+        return this.X.touchManager || this.TouchManager.create();
```
Comment (adamvy) [inline js/foam/browser/u2/BrowserView.js]: Instead of factories like this you can import touchManager, and just leave the factory as return this.TouchManager.create().  The factory won't run if the import sets the value.

File: js/foam/browser/u2/StackView.js [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+          left: this.id$el.offsetWidth,
+          transition: 'left 300ms ease'
+        });
+      } else if (this.transition === 'fade') {
+        e.style({
+          opacity: 0,
+          transition: 'opacity 300ms ease'
+        });
+        // TODO(braden): Probably replace this with requestAnimationFrame?
+        this.setTimeout(function() { e.style({ opacity: 1 }); }, 50);
```
Comment (adamvy) [inline js/foam/browser/u2/StackView.js]: I would just go ahead and replace it with requestAnimationFrame.

File: js/foam/browser/u2/StackView.js [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+          '$$DOC{ref:".popChildViews_"} for details.',
+      code: function() {
+        this.popView_(1);
+      }
+    },
+    function elementAnimationAdd_(e) {
+      if (this.transition === 'slide') {
+        e.style({
+          width: '0px',
+          left: this.id$el.offsetWidth,
```
Comment (adamvy) [inline js/foam/browser/u2/StackView.js]: Do we need to access this.id$el.offsetWidth ?  What about something like left: 100%?  Or just pick a fixed big value.

File: js/foam/browser/u2/StackView.js [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+          opacity: 0,
+          transition: 'opacity 300ms ease'
+        });
+        // TODO(braden): Probably replace this with requestAnimationFrame?
+        this.setTimeout(function() { e.style({ opacity: 1 }); }, 50);
+      }
+    },
+    function elementAnimationRemove_(e) {
+      if (this.transition === 'slide') {
+        e.style({ left: this.id$el.offsetWidth });
```
Comment (adamvy) [inline js/foam/browser/u2/StackView.js]: Ditto

File: js/foam/browser/u2/StackView.js [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+        this.finishDestroy(obj.panel);
+      }
+    },
+    function finishDestroy(e) {
+      this.setTimeout(function() {
+        // Clean up after the animation.
+        // TODO(braden): Not sure if this is done?
+        //obj.view.remove();
+        e.unload();
+        e.destroy();
```
Comment (adamvy) [inline js/foam/browser/u2/StackView.js]: I feel like we should feature request that Element.destroy() calls Element.unload if necessary.

File: js/foam/browser/u2/StackView.js [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+        this.elementAnimationRemove_(obj.panel);
+        this.resize();
+        this.finishDestroy(obj.panel);
+      }
+    },
+    function finishDestroy(e) {
+      this.setTimeout(function() {
+        // Clean up after the animation.
+        // TODO(braden): Not sure if this is done?
+        //obj.view.remove();
```
Comment (adamvy) [inline js/foam/browser/u2/StackView.js]: Shouldn't need this.

File: js/foam/browser/u2/StackView.js [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+      code: function(index, view, hints) {
+        if (!view) return;
+        if (!hints) hints = {};
+
+        // Pushing a new view with index i means we drop everything in the list
+        // that's greater than i, and then add the new view back. There will be
+        // i + 1 views in the stack after a pushView(i, v);
+        this.destroyChildViews_(index + 1);
+
+        var substack = {
```
Comment (adamvy) [inline js/foam/browser/u2/StackView.js]: It would be nice to model substack instead of using __proto__.  This approach will be useful when we want to model the Toolbar object that goes in the context as well.

File: index.js (hunk 8 above)
Comment (shepheb) [inline index.js]: I've added a TODO for now. There doesn't seem to be anything that has a `toE` but isn't an `Element` itself.

File: js/foam/u2/Element.js (hunk 7 above)
Comment (adamvy) [inline js/foam/u2/Element.js]: I disagree.  Element's properties are pretty much just virtual DOM properties.  It's not nearly as polluted as View is.  And even if it was a little poluted I don't think that's justification to just keep stuffing properties in here that we find useful in a few contexts.

Having to worry about a trait isn't that bad.  You  still have to set sensible values for these properties if you want to put it in a stack view, and therefore you're pretty much aware of the trait.

If you don't apply the trait and/or don't set sensible values, then the stack view can pick defaults, rather than every element having defaults that probably don't make any sense.

File: js/foam/u2/Element.js [hunk 17]
```diff
@@ -830,6 +873,7 @@ CLASS({
 
       var first = true;
       for ( var key in this.classes ) {
+        if ( ! this.classes[key] ) continue;
```
Comment (adamvy) [inline js/foam/u2/Element.js]: Seems like an odd check.  when is this happening?

File: js/foam/u2/UpdateView.js (hunk 2 above)
Comment (adamvy) [inline js/foam/u2/UpdateView.js]: Are you planning to remove this for ObjectReplicator in this PR?

We shouldn't be adding throwaway code under foam.u2

File: index.js (hunk 8 above)
Comment (adamvy) [inline index.js]: Element doesn't have toE, or at least it shouldn't.  toE is to adapt something that isn't an Element to an Element.

foam.graphics.CView will probably get a toE soon.

File: js/foam/u2/UpdateView.js (hunk 2 above)
Comment (shepheb) [inline js/foam/u2/UpdateView.js]: This has been replaced in the PR.

File: js/foam/u2/Element.js (hunk 7 above)
Comment (shepheb) [inline js/foam/u2/Element.js]: Done per discussion. See `foam.u2.SizingTrait`.

File: js/foam/browser/BrowserConfig.js (hunk 9 above)
Comment (shepheb) [inline js/foam/browser/BrowserConfig.js]: Fixed.

File: js/foam/browser/u2/BrowserView.js (hunk 10 above)
Comment (shepheb) [inline js/foam/browser/u2/BrowserView.js]: Fixed.

File: js/foam/browser/u2/StackView.js (hunk 11 above)
Comment (shepheb) [inline js/foam/browser/u2/StackView.js]: Done.

File: js/foam/browser/u2/StackView.js (hunk 12 above)
Comment (shepheb) [inline js/foam/browser/u2/StackView.js]: `left: '100%'` seems to work. Done.

File: js/foam/browser/u2/StackView.js (hunk 13 above)
Comment (shepheb) [inline js/foam/browser/u2/StackView.js]: Done.

File: js/foam/browser/u2/StackView.js (hunk 15 above)
Comment (shepheb) [inline js/foam/browser/u2/StackView.js]: Removed.

File: js/foam/browser/u2/StackView.js (hunk 14 above)
Comment (shepheb) [inline js/foam/browser/u2/StackView.js]: Added a TODO.

File: js/foam/browser/u2/StackView.js (hunk 16 above)
Comment (shepheb) [inline js/foam/browser/u2/StackView.js]: The stackview is due for an overhaul as part of this change anyway. I've opened #486 for this.

File: js/foam/u2/Element.js (hunk 17 above)
Comment (shepheb) [inline js/foam/u2/Element.js]: If a class is added using `enableCls('some-class', someValue$, true /* negate */)` then we have `this.classes['some-class'] === false`; but without this check the class was emitted anyway.

Removing classes was working dynamically, but not if they're disabled at output time.

File: js/foam/u2/md/ToolbarAction.js (hunk 3 above)
Comment (adamvy) [inline js/foam/u2/md/ToolbarAction.js]: So is this needed in u2 now?  And when will the new toolbar stuff replace it?

Comment (shepheb) [inline js/foam/u2/md/ToolbarAction.js]: This is still needed. New stuff will be rolling out next week, per @mdittmer .

File: core/mm2Property.js [hunk 18]
```diff
@@ -378,7 +378,16 @@ GLOBAL.Property = {
         e.attrs({size: this.displayWidth});
 
         return e;
-      }
+      },
+      adapt: function(_, nu) {
+        if (typeof nu === 'string') {
```
Comment (kgrgreer) [inline core/mm2Property.js]: I prefer:

return cond ? value1 : value2 ;

When both branches of an if return.

File: js/foam/browser/BrowserConfigU2.js [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+        label: 'Search',
+        floatingLabel: false,
+        placeholder: 'Search',
+        onKeyMode: true,
+      },
+      toPropertyE: function(X) {
+        return X.lookup('foam.u2.md.Input').create({
+          data$: X.data$,
+          inline: true,
+          placeholder: 'Search',
```
Comment (kgrgreer) [inline js/foam/browser/BrowserConfigU2.js]: Should this line be required?

File: js/foam/browser/u2/StackView.js [hunk 20]
```diff
@@ -0,0 +1,412 @@
...
+      }
+      $-no-input {
+        pointer-events: none;
+      }
+    */},
+  ],
+
+  listeners: [
+    {
+      name: 'onLoad',
```
Comment (kgrgreer) [inline js/foam/browser/u2/StackView.js]: Can use shorter form: function onLoad() { ... }

File: js/foam/u2/md/CitationView.js [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+    {
+      name: 'model',
+      defaultValueFn: function() {
+        return this.data.model_;
+      }
+    },
+  ],
+
+  methods: [
+    function pickNameProperty() {
```
Comment (kgrgreer) [inline js/foam/u2/md/CitationView.js]: This looks expensive.  Can we cache this and then invalidate when data is updated?

File: core/mm2Property.js (hunk 18 above)
Comment (shepheb) [inline core/mm2Property.js]: Done.

File: js/foam/browser/BrowserConfigU2.js (hunk 19 above)
Comment (shepheb) [inline js/foam/browser/BrowserConfigU2.js]: Nope. `placeholder` now defaults to `label`.

File: js/foam/browser/u2/StackView.js (hunk 20 above)
Comment (shepheb) [inline js/foam/browser/u2/StackView.js]: DOne.

File: js/foam/u2/md/CitationView.js (hunk 21 above)
Comment (shepheb) [inline js/foam/u2/md/CitationView.js]: Now cached and only updated when the `model` changes.
